### ISSUE TITLE
3.7.9 does not restore DNS settings properly if computer is shut down or restarted while VPN is connected

### ISSUE THREAD

**Issue by kcshiu** [Tunnelblick/Tunnelblick#539]
**Describe the bug**
Running on MacOS 10.14.5 (18F132)
This version will set the dns on the client machine to routed dns upon vpn connection.
The dns remains upon disconnect. A reconnection will use this dns which is now unavailable
. Manually reseting the dns to a local address will solve this. This pattern repeats with each vpn connection.

**To Reproduce**
1. Connect to a vpn server that set dns on the remote site
2. Disconnect and reconnect the vpn. It should fail because the dns are no longer available.
3. Check the local dns address and reset. The vpn will reconnect again

**Expected behavior**
This behaviour was not present in the previous version

**Screenshots**

**Additional context**

**Diagnostic information**
Please paste or attach the diagnostic info obtained by following the instructions at [Read Before You Post](https://tunnelblick.net/cBeforeYouPost.html) here.

**Comment by jkbullard** [comment]
This is not a common problem; it has to do with your specific setup.

Please paste or attach the diagnostic info obtained by following the instructions at [Read Before You Post](https://tunnelblick.net/cBeforeYouPost.html) here.

**Comment by jozsi** [comment]
I don't think it's a specific issue, I've been using the same setup with Tunneblick for years and this issue appeared after the latest update (build 5320).

**Comment by jkbullard** [comment]
Here's the problem:

When the VPN connects, the log shows:
       *Tunnelblick: Did not change DNS ServerAddresses setting of '10.147.220.138 10.9.190.103' (but re-set it)
...
       *Tunnelblick: DNS servers '10.147.220.138 10.9.190.103' were set manually

Please remove the manual setting, which is to the VPN's DNS. Then get the diagnostic info again.

Note: You apparently have (or had) other VPN software (you have the tuntaposx kexts loaded); that software could have changed the DNS servers, but in any case you need to remove the manual entry and get new diagnostic info.

**Comment by jozsi** [comment]
Yes, I did have an attempt to connect to an academic library using a CLI tool, but that was months ago and haven't used it since (+ it didn't touch the nameservers).

I manually removed the nameservers in the Advanced tab of the Network settings. 
![image](https://user-images.githubusercontent.com/840815/58345887-21c58b80-7e62-11e9-9824-a43f020f88db.png)

Connecting and disconnecting to the VPN now resets the nameservers correctly.

I do beleive there was a bug during the update process.

**Comment by jkbullard** [comment]
OK, thanks for letting us know.

The only way I can imagine the manual settings getting "stuck" is if the update was done while the VPN was connected. The update process closes Tunnelblick but maybe it doesn't wait for Tunnelblick to actually close before continuing.

I've changed this Issue's title to reflect that (possible) problem and will leave it "open" until that is resolved.

**Comment by jozsi** [comment]
I am pretty sure I updated on system startup, when I got the notification. Before being connected.

**Comment by jkbullard** [comment]
@jozsi - OK, thanks. I'll look into the possibility in any case.

**Comment by GuillotJessica** [comment]
I have the same problem, i fixed it the first time with removing dns servers and dns domains list and the next time by renewing DHCP Lease, i'm using Tunnelblick 3.7.9 and mojave 10.14.4, i didn't have the issue before i upgraded tunnelblick

**Comment by jozsi** [comment]
It happened the second time for me, so it's not during the update but in the new version itself. I just normally shut down the laptop while being connected, on next boot I had the VPN-set dns servers in `/etc/resolv.conf` (prior to connect to it).

**Comment by Anrock** [comment]
Same here. Every time I shutdown my mac with vpn connection active on next startup it has vpn dnses in network settings and each time i have to remove it manually.

**Comment by jkbullard** [comment]
It would be very helpful to know if this ever happens when the VPN is disconnected _before_ shutting down.

**Comment by Anrock** [comment]
I'll check this tomorrow.

**Comment by mdugue** [comment]
same here

**Comment by Anrock** [comment]
@jkbullard everything is fine, if VPN disconnected prior to shutdown.

**Comment by jkbullard** [comment]
@Anrock - Thanks! That helps narrow down the problem.

Additional confirmation of this would be great, since it may be different for different users.

**Comment by Anrock** [comment]
@jkbullard IMO, it is the same for everyone. Disconnecting VPN clears DNS as expected and it doesn't matter if shutdown will or will not happen after it. 
My wild guess is that when OS closes active apps during shutdown Tunnelblick either not fast enough to disconnect and clear DNS or it doesn't ask OS to postpone shutdown until everything VPN-related is taken care of.

**Comment by jkbullard** [comment]
I would agree with @Anrock that
> when OS closes active apps during shutdown Tunnelblick either not fast enough to disconnect and clear DNS or it doesn't ask OS to postpone shutdown until everything VPN-related is taken care of.

except for the log from [1]. It's the first and only log I have seen that shows the problem as it happened. It shows

(1) A normal connection, with the 'up' script saving the DNS (and other) info.

(2) Normal things happening while the VPN is connected (two irrelevant system configuration changes were ignored, a domain name change (reversed by Tunnelblick's "process-network-changes" subprogram), which was probably as a result of a DHCP renewal, and a key renegotiation sequence by OpenVPN:
```
2019-05-26 08:59:40.027795 *Tunnelblick: process-network-changes: A system configuration change was ignored
2019-05-26 09:28:53.763467 *Tunnelblick: process-network-changes: DomainName changed from
*openvpn
* to (pre-VPN)
*domain
2019-05-26 09:28:53.763910 *Tunnelblick: process-network-changes: Restoring DomainName to post-VPN value
2019-05-26 09:29:04.212258 *Tunnelblick: process-network-changes: A system configuration change was ignored
2019-05-26 09:59:29.000408 TLS: soft reset sec=0 bytes=1173931173/-1 pkts=1043997/0
.
.
.
2019-05-26 10:59:29.708295 Control Channel: TLSv1.2, cipher TLSv1/SSLv3 DHE-RSA-AES256-SHA, 2048 bit RSA
```

(3) A disconnection but the saved info has disappeared:

```
2019-05-26 11:19:32.176429 event_wait : Interrupted system call (code=4)
.
.
.
*Tunnelblick: **********************************************
*Tunnelblick: Start of output from client.down.tunnelblick.sh
*Tunnelblick: WARNING: Not restoring DNS settings because no saved Tunnelblick DNS information was found.
```
The info disappeared without any shutdown or restart, or anything unusual at all!


[1] The first attachment, named "2019-05-26.tunnelblick.log1.warning.about.no.saved.DNS.info.found.redacted.txt",  from a post at https://groups.google.com/d/msg/tunnelblick-discuss/55b_omU6upE/TbFB6_lQBwAJ

**Comment by Anrock** [comment]
I have another interesting bit. My VPN config uses some deprecated parameters, so Tunnelblick shows a dialog window with warning every time I connect/disconnected VPN. Is it possible that this window blocks some processes inside Tunnelblick during the shutdown?

**Comment by jkbullard** [comment]
@Anrock wrote
> Tunnelblick shows a dialog window with warning every time I connect/disconnected VPN. Is it possible that this window blocks some processes inside Tunnelblick during the shutdown?

When the window is displayed, if you click the Tunnelblick icon, does it respond? If it does, then nothing important is being blocked.

**Comment by Anrock** [comment]
@jkbullard AFAIK it does respond.

**Comment by jozsi** [comment]
I did manage to reproduce this once with 3.7.9a as well.

**Comment by polak785** [comment]
> I did manage to reproduce this once with 3.7.9a as well.

Here too. We have around 40 users and 2 had the same issue with version up-to-date including the fix (3.7.9a). Moreover, at least for one of it, the issue came after a sleep and not a restart/shutdown.

**Comment by wosc** [comment]
I still have this issue regularly with 3.7.9a: VPN is connected, shut down mac, start mac: VPN-set DNS servers are still there and have to be removed manually.

**Comment by jkbullard** [comment]
We expect will be fixed in the next beta release of Tunnelblick, due soon.

Until it is fixed, you can avoid the problem by disconnecting from all VPNs before you shut down or restart your computer. It isn't necessary to quit Tunnelblick.

**Comment by jkbullard** [comment]
This is fixed in Tunnelblick 3.8.0beta03.

**Comment by wosc** [comment]
Thanks for the update! And yeah, I've been doing this, disconnecting or quitting before shutdown (if I remember to do it ;) to avoid the issue on the next boot.
